pybdv/bdv_datasets.py

```diff
@@ -0,0 +1,213 @@
+            in_target_pos[2]: in_target_pos[2] + volume.shape[2]
+        ] = volume
+
+    else:
+
+        # Everything scales nicely, so we just have to take care that the proper variables exist
+        target_vol = volume
+        target_pos = position
+        target_shape = vol_shape
+
+    return target_vol, target_pos, target_shape
+
+
+def _scale_and_add_to_dataset(
+        data_path, setup_id, timepoint,
+        target_pos, target_vol, target_shape,
+        scales, downscale_mode, n_threads):
+
+    # Perform scaling of volume, by generating a temporary bdv file
+    # FIXME it is probably sufficient to keep it in memory, since I assume that the volume is in memory anyways, but the
+    # FIXME     make_bdv function is so very convenient right now...
+    scale_factors = scales[1: 2].tolist()
+    tmp_name = os.path.join(data_path, f'tmp_{np.random.randint(0, 2 ** 16, dtype="uint16")}')
+    tmp_bdv = tmp_name + '.n5'
+    for scale in scales[2:]:
+        scale_factors.append((scale / np.product(scale_factors, axis=0)).astype(int).tolist())
+    make_bdv(
+        data=target_vol,
+        output_path=tmp_bdv,
+        downscale_factors=scale_factors,
+        downscale_mode=downscale_mode,
+        n_threads=n_threads
+    )
+
+    # Now, we just need to fetch the scaled data and put it to the proper positions
+    for scale_id, scale in enumerate(scales):
+
+        # Position in the current scale
+        pos_in_scale = (target_pos / scale).astype(int)
+        shp_in_scale = (target_shape / scale).astype(int)
+
+        # Write the data to each scale
+        with open_file(tmp_bdv, mode='r') as f:
+            scaled_vol = f[f'setup0/timepoint0/s{scale_id}'][:]
+            # assert list(scaled_vol.shape) == shp_in_scale.tolist()
+
+        with open_file(data_path, mode='a') as f:
+            f[f'setup{setup_id}/timepoint{timepoint}/s{scale_id}'][
+                pos_in_scale[0]: pos_in_scale[0] + shp_in_scale[0],
+                pos_in_scale[1]: pos_in_scale[1] + shp_in_scale[1],
+                pos_in_scale[2]: pos_in_scale[2] + shp_in_scale[2]
+            ] = scaled_vol
+
+    # Delete the temporary bdv file
+    rmtree(tmp_bdv)
+    os.remove(tmp_name + '.xml')
+
+
+class BdvDataset:
+    """
+    The basic BDV dataset to which volumes can be written using numpy nomenclature.
+
+    The data is included into each of the down-sampling layers accordingly
+    The full resolution area is padded, if necessary, to avoid sub-pixel locations in the down-sampling layers
+    """
+
+    def __init__(self, path, timepoint, setup_id, n_threads=1, verbose=False):
+
+        self._path = path
+        self._timepoint = timepoint
+        self._setup_id = setup_id
+        self._n_threads = n_threads
+        self._verbose = verbose
+
+    def _add_to_volume(self, position, volume):
+
+        # Determine required scales
+        with open(os.path.join(self._path, 'setup{}'.format(self._setup_id), 'attributes.json')) as f:
```

Comment:
Here you assume that we have an n5 dataset, but I think it would be nice to make it work for h5 as well.
I actually have a helper function for this already.
https://github.com/constantinpape/pybdv/blob/master/pybdv/util.py#L139

Also, you could do this call in the init and have `scales` as a member variables, because this is not expected to change.